### PR TITLE
Feat: support unsafe-eval

### PR DESCRIPTION
**Pull request by achamas-playco**
### Summary
🛡️🎨🧰

This pull request adds a new `unsafe-eval` module that polyfills the use of eval and new Function in PixiJS for CSP-restricted environments. It also refactors and improves the code for generating and syncing the uniforms and the uniform buffers of shader programs, by adding type annotations, constants, types, and utility functions. It also updates the `package.json` file to export the new module as a subpath and a conditional export.

> _Sing, O Muse, of the cunning code that syncs the uniforms_
> _Of the shader programs that render the glorious scenes of PixiJS,_
> _The code that uses `eval` and `new Function` to create dynamic functions,_
> _But also provides a polyfill for the browsers that shun these methods._

### Walkthrough
*  Add a new unsafe-eval module that provides a polyfill for environments that do not support unsafe evals, such as CSP-restricted browsers ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L137-R138), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R246-R253), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-696d1b23b9b8ac4c207130e7614bd38daa4d92f03c23ae32c1068aa4bd09f09bL24-R24), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-696d1b23b9b8ac4c207130e7614bd38daa4d92f03c23ae32c1068aa4bd09f09bL65-R65), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-070716c749065580d29476152682d62d0e29f7a00e0d8c784e3cca3033edecdaR1-R33))
*  Modify the generateUniformsSync and the generateUniformBufferSync functions that create dynamic functions to sync the uniforms and the uniform buffers of shader programs, to use the _generateUniformsSync and the _generateUniformBufferSync methods of the GlUniformGroupSystem and the UniformBufferSystem classes, which can be overridden by the unsafe-eval module to provide a polyfill ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-0a4c66cbfc80c26a635ff3ac6d798748f4a6a3b464de2d0d65aae76269a02514L109-R109), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-0a4c66cbfc80c26a635ff3ac6d798748f4a6a3b464de2d0d65aae76269a02514R117-R121), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-32bd5aafe934ed5702790d3c8f090b3d7de0daeebdd3250e70001b38a7d40af4L50-R71), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-32bd5aafe934ed5702790d3c8f090b3d7de0daeebdd3250e70001b38a7d40af4R89-R95))
*  Add the _systemCheck method to the UniformBufferSystem class that checks if the current environment supports unsafe evals and throws an error if not, unless the unsafe-eval module is installed to provide a polyfill ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-32bd5aafe934ed5702790d3c8f090b3d7de0daeebdd3250e70001b38a7d40af4R2), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-32bd5aafe934ed5702790d3c8f090b3d7de0daeebdd3250e70001b38a7d40af4R25-R44))
*  Add the exec property to the uniformParsers and the uniformBufferParsers arrays that defines custom logic for setting different types of uniforms and uniform buffer elements, and use the parseFunctionBody function to extract the body of the exec function as a string for the code property ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-a7b0bfdce40fc7d05963936597a1ff8eaa9d66b1473c4731a269b3f8e7791ab2L13-R236), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-847ff44999e66423fee8a46677b0624d1f9ae449d2e9fb77ec50877e130706dfL22-R113), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-f2dbd6ee1e816d3c6c0454ee62bd4e3fd91912f97aee2afed37d3ecd424ecaaaR1-R13))
*  Add the UBO_TO_SINGLE_SETTERS_FN and the GLSL_TO_SINGLE_SETTERS_FN_CACHED constants that implement the logic for setting different types of WGSL and GLSL uniforms, and use the parseFunctionBody function to derive the UBO_TO_SINGLE_SETTERS and the GLSL_TO_SINGLE_SETTERS_CACHED constants that contain the code for setting different types of WGSL and GLSL uniforms ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-6fe64c9e77c4938a4fe2235714921ffdd700a5ab17ff57f7737d6e08faf52f92R1-R98), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-6b2036437a03a35fd4d692b569fb975c0379b8c5a6a304703a3a9bfa68a1e834L9-R14))
*  Add the name variable to the generateUniformsSync and the generateUniformBufferSync functions that stores the name of the current uniform or uniform buffer element being processed, and use it by the uniformParsers and the uniformBufferParsers to access the uniform value from the uniform group object, instead of hardcoding the name in the code template ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-6b2036437a03a35fd4d692b569fb975c0379b8c5a6a304703a3a9bfa68a1e834R23), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-6b2036437a03a35fd4d692b569fb975c0379b8c5a6a304703a3a9bfa68a1e834L265-R65), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-20bc4079319d916453e58673fff57610c247ffbbfc27e968f90d98c113b7c427R18), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-20bc4079319d916453e58673fff57610c247ffbbfc27e968f90d98c113b7c427L114-R49))
*  Add a block of code to the generateUniformBufferSync function that tests and executes a dummy code snippet to ensure that the generated function has a valid syntax and does not throw an error when using new Function or eval, as a workaround for a bug in Safari ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-20bc4079319d916453e58673fff57610c247ffbbfc27e968f90d98c113b7c427R32-R37))
*  Add the type annotations to the uniformData parameter of the _generateUniformsSync method and the parameters and return value of the _generateUniformBufferSync method, which were previously untyped ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-0a4c66cbfc80c26a635ff3ac6d798748f4a6a3b464de2d0d65aae76269a02514L6-R10), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-32bd5aafe934ed5702790d3c8f090b3d7de0daeebdd3250e70001b38a7d40af4L9-R11))
*  Cast the data.type property of the uniforms and the uboElement.data.type property of the uniform buffer elements to the GLSL_TYPE and the UBO_TYPE types, respectively, to avoid TypeScript errors when using them as keys for the template objects ([link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-6b2036437a03a35fd4d692b569fb975c0379b8c5a6a304703a3a9bfa68a1e834L275-R75), [link](https://github.com/pixijs/pixijs/pull/9778/files?diff=unified&w=0#diff-20bc4079319d916453e58673fff57610c247ffbbfc27e968f90d98c113b7c427L153-R87))

### CODE REVIEW

File: src/rendering/renderers/gl/shader/program/generateUniformsSync.ts [hunk 1]
```diff
@@ -262,7 +62,8 @@ export function generateUniformsSync(group: UniformGroup, uniformData: Record<st
         {
             if (uniformParsers[j].test(data, uniform))
             {
-                funcFragments.push(uniformParsers[j].code(i, uniform));
+                funcFragments.push(`name = "${i}";`);
+                funcFragments.push(uniformParsers[j].code);
```
Comment (GoodBoyDigital) [inline src/rendering/renderers/gl/shader/program/generateUniformsSync.ts:66]: could do it in one push!
```
funcFragments.push(
   `name = "${i}";`
    uniformParsers[j].code
)
```

File: src/unsafe-eval/init.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    Object.assign(GlUniformGroupSystem.prototype,
+        {
+            _systemCheck()
+            {
+                // Do nothing, don't throw error
+            },
+
+            _generateUniformsSync(group: UniformGroup, uniformData: Record<string, IUniformData>): UniformsSyncCallback
+            {
+                return generateUniformsSyncPolyfill(group, uniformData);
```
Comment (GoodBoyDigital) [inline src/unsafe-eval/init.ts:19]: you don't need to generate this each time as the the single polyfill function should cover every shader. You can basically return the same function every time here!

**Review by GoodBoyDigital — COMMENTED** [review]
Looks good! Love this PR. Only tweak is that we don't need to return a new polyfill function for different shaders. The same function will work every time on every shader.

if we can tweak that, we are good to go!

File: src/unsafe-eval/init.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { GlUniformGroupSystem } from '../rendering/renderers/gl/shader/GlUniformGroupSystem';
+import { generateUniformsSyncPolyfill } from './generateUniformsSyncPolyfill';
+
+import type { IUniformData } from '../rendering/renderers/gl/shader/GlProgram';
+import type { UniformGroup } from '../rendering/renderers/shared/shader/UniformGroup';
+import type { UniformsSyncCallback } from '../rendering/renderers/shared/shader/utils/createUniformBufferSync';
+
+export function selfInstall()
```
Comment (GoodBoyDigital) [inline src/unsafe-eval/init.ts:8]:
```suggestion function selfInstall()
```

Comment (GoodBoyDigital) [inline src/unsafe-eval/init.ts:8]:
```suggestion
function selfInstall()
```

**Comment by achamas-playco** [comment]
> Looks good! Love this PR. Only tweak is that we don't need to return a new polyfill function for different shaders. The same function will work every time on every shader.
> 
> if we can tweak that, we are good to go!

Thanks! I've added the uniform buffer changes and polyfill similar to the uniform sync work. Would be great if you could QA it locally with any specific use cases or paths which would functionally verify it. Hopefully all the logic was preserved emulating the generated source. I've tested with the playground with the unsafe-eval package on/off and so far so good.

**Review by GoodBoyDigital — APPROVED** [review]
looking good!! nice work @achamas-playco 👍

**Pull request merged**